### PR TITLE
Add migration guide from ink! 4.x to 5.0

### PR DESCRIPTION
**Pull request by cmichi**
Most static image files are already in `master`.

### CODE REVIEW

**Review by smiasojed — APPROVED** [review]

File: versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+security review](https://blog.openzeppelin.com/security-review-ink-cargo-contract).
+In particular, we addressed the proxy selector clashing attack.
+
+You can find the full changelog of the 5.0 release [here](https://github.com/paritytech/ink/blob/master/CHANGELOG.md).
+
+:::caution
+This migration guide only considers your code base! Not your data!
+
+If you have an existing contract on-chain you cannot just
+upgrade the code on-chain, you also have to migrate your data.
```
Comment (SkymanOne) [inline versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md:22]: Same as earlier

File: versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+
+The way how ink! 4.0 stores data and reads from it (i.e. the storage
+layout) changes from ink! 3.x to 4.0.
```
Comment (SkymanOne) [inline versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md:25]: How is this related to migration from 4.x to 5.0?

File: versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+
+Following [our security review by OpenZeppelin](https://blog.openzeppelin.com/security-review-ink-cargo-contract),
+we've tightened the usage of wildcard selectors.
+With ink! 5.0 we allow only exactly one other contract message with a well-known reserved
+selector to be defined. In ink! 4.x, more than one other message was allowed.
+
+Read more in [the PR](https://github.com/paritytech/ink/pull/1708) and [IIP-2: Limit contracts with a wildcard selector to one other message](https://github.com/paritytech/ink/issues/1676).
+
+The proposal is to restrict contracts with a wildcard selector to only have one other message
+with a reserved/well known selector. This guarantees that there are no selector clashes,
```
Comment (SkymanOne) [inline versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md:181]:
```suggestion
with a reserved/well-known selector. This guarantees that there are no selector clashes,
```

File: versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+The syntax of defining events within the main `#[ink::contract]` macro will continue to work,
+no code changes in existing contracts are required to update to the new syntax.
+
+:::caution
+The topic calculation changed in general, so also for events that are declared inside the
+`#[ink::contract]` macro!
+
+This is a breaking change for any client code which uses topics to filter events.
+
+Please see [#1827](https://github.com/paritytech/ink/pull/1827) for details.
```
Comment (SkymanOne) [inline versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md:234]: Need also to include https://github.com/paritytech/ink/pull/2031

File: versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+:::
+
+### No more unchecked arithmetic
+
+Unchecked arithmetic operations in a contract no longer work.
```
Comment (SkymanOne) [inline versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md:239]:
```suggestion
Unchecked arithmetic operations in a contract are no longer supported for arithmetic safety reasons.
```

File: versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+then recompiling with a version of ink! which includes this change will cause that contract
+to trap at runtime when attempting to decode.
+
+A simple example would be if a storage cell contains some bytes which were in the first place
+an encoded `u32`. If the contract attempts to decode those into a `u8`
+this would previously have succeeded, now the contract would trap.
+
+We introduced this change in [#1897](https://github.com/paritytech/ink/pull/1897).
+
+### [ink_e2e] Improved `call()` API
```
Comment (SkymanOne) [inline versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md:257]: Missing the notion of builder API in e2e: https://github.com/paritytech/ink/pull/1917

File: versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+between `pallet-contracts` and a smart contract. The advantage of a static buffer
+is that no gas-expensive heap allocations are necessary, all allocations are done
+using simple pointer arithmetic.
+
+The default static buffer size is 16 kB, which is enough for on-chain smart
+contracts. However, the [Phala Network](https://phala.network/) parachain on Polkadot
+allows the deployment of ink! contracts off-chain. Hence, for their chain certain high
+computation contracts might require a larger buffer size.
+
+### Custom signature topics
```
Comment (SkymanOne) [inline versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md:445]: I think it may be moved as a subsection of Events 2.0

File: versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+Please see [#1827](https://github.com/paritytech/ink/pull/1827) for details.
+:::
+
+### No more unchecked arithmetic
+
+Unchecked arithmetic operations in a contract no longer work.
+Compiling a contract that contains those will fail gracefully.
+You need to handle overflows using e.g. saturating or wraparound math.
+
+This was done in [#1831](https://github.com/paritytech/ink/pull/1831).
```
Comment (xermicus) [inline versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md:243]: The guide could briefly explain the alternatives (`unchecked`, `saturating`) and how to use them

File: versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+used during ABI encoding and decoding (default 16 KiB).
+`StorageVec` on the other hand allows to access each element individually.
+
+With a `Vec` it's possible to e.g. introduce a security issue in your contract
+where an attacker can fill the `Vec`, making it very costly for other users to
+access it or write to it.
+
+You can find verbatim documentation on `StorageVec` [here](/5.x/datastructures/storagevec).
+The page explains when to use `StorageVec` and when not.
+The Rust docs can be found [here](https://docs.rs/ink/5.0.0-rc/ink/storage/struct.StorageVec.html).
```
Comment (xermicus) [inline versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md:291]: LGTM!

File: versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+* [`Lazy`](https://docs.rs/ink/5.0.0-rc/ink/storage/struct.Lazy.html)
+* [`Mapping`](https://docs.rs/ink/5.0.0-rc/ink/storage/struct.Mapping.html). For `Mapping`, the encoded size of the key is also accounted for.
+
+You should use the `try_*` methods for dynamically sized values, unless you made sure
+otherwise they will fit into the static buffer. The [static buffer in ink!](https://github.com/paritytech/ink/blob/master/ARCHITECTURE.md#communication-with-the-pallet)
+is 16 kB by default.
+
+We added a lint to `cargo-contract` 4.0 that will detect
+potentially unsafe uses of methods for which there are safer alternatives:
+[`non_fallible_api`](https://use.ink/5.x/linter/rules/non_fallible_api).
```
Comment (xermicus) [inline versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md:313]: LGTM!

File: versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md (hunk 2 above)
Comment (cmichi) [reply versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md:25]: Thanks, improved it.

File: versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md (hunk 8 above)
Comment (SkymanOne) [reply versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md]: I would just provide a link to the rust docs.
